cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -153,7 +188,32 @@ protected String processMessages(Iterable<ImmutablePair<String, T>> messages) th
       return trackingIterator;
     }, Exception.class);
 
-    return iterator.getLastMessageId();
+    return iterator.lastMessageId;
+  }
+
+  private final class BatchIterator extends AbstractIterator<List> {
+    private final MessageTrackingIterator delegate;
+    private final int batchSize;
+
+    private List<ImmutablePair<String, T>> currentBatch;
```

Comment:
this will result in an extra copy of the data for the default case where fetch size = tx size. 

Instead of doing it this way, we could have a different type of wrapper, something like a LimitIterator, which returns up to n elements. Then create a new LimitIterator for each batch.

Reply:
Added a case to bypass batching if txSize is greater or equal to fetchSize.